File: app/src/main/java/com/chengsheng/cala/htcm/widget/AppTitleBar.java
package com.chengsheng.cala.htcm.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.chengsheng.cala.htcm.R;

/**
 * Author: 任和
 * CreateDate: 2018/12/17 6:10 PM
 * Description: APP通用标题栏
 */
public class AppTitleBar extends FrameLayout {
    /**
     * 页面标题
     */
    private TextView tvTitle;
    /**
     * 右边可操作图标
     */
    private ImageView ivRightAction;
    /**
     * 右边可操作文字
     */
    private TextView tvRightAction;

    private OnViewClickListener onFinishClickListener;
    private OnViewClickListener onRightImageClickListener;
    private OnViewClickListener onRightTxtClickListener;

    public AppTitleBar(Context context) {
        super(context);
    }

    public AppTitleBar(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public AppTitleBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        View contentView = LayoutInflater.from(context).inflate(R.layout.layout_titlebar, this);
        ImageView ivFinishPage = contentView.findViewById(R.id.ivFinishPager);
        tvTitle = contentView.findViewById(R.id.tvPagerTitle);
        ivRightAction = contentView.findViewById(R.id.ivRightImageAction);
        tvRightAction = contentView.findViewById(R.id.tvRightAction);
        View viewDivide = contentView.findViewById(R.id.viewDivide);

        //获取自定义属性
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.titleBar);
        String title = array.getString(R.styleable.titleBar_title);
        Boolean isShowFinish = array.getBoolean(R.styleable.titleBar_isShowReturn, true);
        Boolean isShowDivide = array.getBoolean(R.styleable.titleBar_isShowDivide, true);
        int rightImgResId = array.getResourceId(R.styleable.titleBar_rightImageRes, -1);
        String rightTxt = array.getString(R.styleable.titleBar_rightTxtString);
        int rightTxtColor = array.getResourceId(R.styleable.titleBar_rightTxtColor, -1);
        array.recycle();

        /*
         * 返回按钮相关
         */
        if (ivFinishPage != null) {
            //设置是否显示返回按钮
            ivFinishPage.setVisibility(isShowFinish ? View.VISIBLE : View.INVISIBLE);
            //设置返回事件的点击事件
            ivFinishPage.setOnClickListener(view -> {
                if (onFinishClickListener != null) {
                    onFinishClickListener.onClick();
                }
            });
        }

        //设置页面标题
        if (tvTitle != null) {
            tvTitle.setText(title);
        }

        /*
         * 右边图标相关
         */
        if (ivRightAction != null) {
            //设置(-1为默认值，不显示图标)
            if (rightImgResId != -1) {
                ivRightAction.setVisibility(View.VISIBLE);
                ivRightAction.setImageResource(rightImgResId);
            }

            ivRightAction.setOnClickListener(view -> {
                if (onRightImageClickListener != null) {
                    onRightImageClickListener.onClick();
                }
            });
        }

        //设置是否显示分割线
        viewDivide.setVisibility(isShowDivide ? View.VISIBLE : View.INVISIBLE);

        /*
         * 右边文字相关
         */
        if (tvRightAction != null) {
            tvRightAction.setOnClickListener(view -> {
                        if (onRightTxtClickListener != null) {
                            onRightTxtClickListener.onClick();
                        }
                    }
            );

            //文字文本
            if (rightTxt != null) {
                tvRightAction.setVisibility(View.VISIBLE);
                tvRightAction.setText(rightTxt);
            }

            //右边文字颜色(颜色一定要抽取到color.xml下)
            if (rightTxtColor != -1) {
                tvRightAction.setVisibility(View.VISIBLE);
                tvRightAction.setTextColor(getResources().getColor(rightTxtColor));
            }
        }
    }

    /**
     * 设置页面标题
     */
    public AppTitleBar setTitle(String title) {
        tvTitle.setText(title);
        return this;
    }

    /**
     * 设置返回事件
     */
    public AppTitleBar setFinishClickListener(OnViewClickListener clickListener) {
        this.onFinishClickListener = clickListener;
        return this;
    }

    /**
     * 设置右边图标点击事件
     */
    public AppTitleBar setRightImageClickListener(OnViewClickListener clickListener) {
        this.onRightImageClickListener = clickListener;
        return this;
    }

    /**
     * 设置右边文字点击事件
     */
    public AppTitleBar setRightTxtClickListener(OnViewClickListener clickListener) {
        this.onRightTxtClickListener = clickListener;
        return this;
    }
}

File: app/src/main/java/com/chengsheng/cala/htcm/widget/AppTitleBar.kt
package com.chengsheng.cala.htcm.widget

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import com.chengsheng.cala.htcm.R

/**
 * Author: 任和
 * CreateDate: 2018/12/17 4:24 PM
 * Description: APP顶部标题栏
 */
class AppTitleBar : FrameLayout {

    private var ivFinishPage: ImageView? = null
    private var tvPageTitle: TextView? = null
    private var ivRightAction: ImageView? = null

    private var rightActionClickListener: OnViewClickListener? = null
    private var finishClickListener: OnViewClickListener? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        initView(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initView(context, attrs)
    }

    private fun initView(context: Context, attrs: AttributeSet?) {
        val contentView = LayoutInflater.from(context).inflate(R.layout.layout_titlebar, this)
        ivFinishPage = contentView.findViewById<View>(R.id.iv_finish_pager) as ImageView
        tvPageTitle = contentView.findViewById<View>(R.id.tv_pager_title) as TextView
        ivRightAction = contentView.findViewById(R.id.iv_right_action) as ImageView

        //获取自定义的属性
        val array = context.obtainStyledAttributes(attrs, R.styleable.titleBar)
        val title = array.getString(R.styleable.titleBar_title)

        val isShowFinish = array.getBoolean(R.styleable.titleBar_isShowReturn, true)
        val rightImgResId: Int = array.getResourceId(R.styleable.titleBar_rightImageView, -1)
        array.recycle()

        if (rightImgResId != -1) {
            ivRightAction?.setImageResource(rightImgResId)
        } else {
            ivRightAction?.visibility = View.INVISIBLE
        }

        //设置返回键是否显示
        ivFinishPage?.visibility = if (isShowFinish) View.VISIBLE else View.INVISIBLE

        //设置默认标题
        if (tvPageTitle != null && !TextUtils.isEmpty(title)) {
            tvPageTitle?.text = title
        }

        //返回事件
        ivFinishPage?.setOnClickListener { v ->
            if (finishClickListener != null) {
                finishClickListener?.onClick()
            }
        }

        //右边操作事件
        ivRightAction?.setOnClickListener {
            if (rightActionClickListener != null) {
                rightActionClickListener?.onClick()
            }
        }
    }

    /**
     * 设置标题
     */
    fun setTitle(title: String): AppTitleBar {
        if (tvPageTitle != null) {
            tvPageTitle?.text = title
        }
        return this
    }

    /**
     * 设置返回点击事件
     */
    fun setOnFinishClickListener(returnClickListener: OnViewClickListener): AppTitleBar {
        this.finishClickListener = returnClickListener
        return this
    }

    /**
     * 设置右边操作按钮点击事件
     */
    fun setOnRightClickListener(rightActionClickListener: OnViewClickListener): AppTitleBar {
        this.rightActionClickListener = rightActionClickListener
        return this
    }

    interface OnViewClickListener {
        fun onClick()
    }
}
